File: android/src/main/kotlin/it/mainella/audible_mode/handler/FlutterHandler.kt
package it.mainella.audible_mode.handler

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import it.mainella.audible_mode.utils.Constants

class FlutterHandler(binding: FlutterPlugin.FlutterPluginBinding) {
    private var channel: MethodChannel = MethodChannel(binding.binaryMessenger, Constants.METHOD_CHANNEL)
    private var currentProfileEventChannel: EventChannel = EventChannel(binding.binaryMessenger, Constants.CURRENT_PROFILE_EVENT)
    private var currentVolumeEventChannel: EventChannel = EventChannel(binding.binaryMessenger, Constants.CURRENT_VOLUME_EVENT)
    private var audibleHandler = AudibleHandler(binding.applicationContext)

    init {
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                Constants.GET_AUDIBLE_MODE -> {
                    result.success(audibleHandler.getAudibleProfile().rawValue)
                }

                Constants.GET_CURRENT_VOLUME -> {
                    result.success(audibleHandler.getCurrentVolume())
                }

                Constants.SET_VOLUME -> {
                    audibleHandler.setVolume(call.argument<Double>("volume"))
                    result.success(true)
                }

                Constants.GET_MAX_VOLUME -> {
                    result.success(audibleHandler.getMaxVolume())
                }

                else -> result.notImplemented()
            }
        }
        currentVolumeEventChannel.setStreamHandler(
            CurrentVolumeStreamHandler(
                context = binding.applicationContext
            )
        )

        currentProfileEventChannel.setStreamHandler(
            CurrentProfileStreamHandler(
                context = binding.applicationContext,
                audibleHandler = audibleHandler
            )
        )
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        currentProfileEventChannel.setStreamHandler(null)
        currentVolumeEventChannel.setStreamHandler(null)
    }
}

File: ios/Classes/handler/FlutterHandler.swift
//
//  FlutterHandler.swift
//  audible_mode
//
//  Created by Andrea Mainella on 22/12/21.
//

import Foundation


public class FlutterHandler{
    
    private let channel: FlutterMethodChannel
    private let currentProfileEventChannel: FlutterEventChannel
    private let currentVolumeEventChannel: FlutterEventChannel
    
    private let _currentVolumeHandler: CurrentVolumeStreamHandler
    
    init(binding: FlutterPluginRegistrar) {
        let audibleHandler: AudibleHandler = AudibleHandler();
        
        channel = FlutterMethodChannel(
            name: Constants.METHOD_CHANNEL,
            binaryMessenger: binding.messenger());
        
        currentProfileEventChannel = FlutterEventChannel(
            name: Constants.CURRENT_PROFILE_EVENT,
            binaryMessenger: binding.messenger());
        
        currentVolumeEventChannel = FlutterEventChannel(
            name: Constants.CURRENT_VOLUME_EVENT,
            binaryMessenger: binding.messenger());
        
        
        channel.setMethodCallHandler{(call: FlutterMethodCall, result: FlutterResult) -> Void in
            switch call.method {
            case Constants.GET_AUDIBLE_MODE:
                result(audibleHandler.getAudibleProfile())
                break;
            case Constants.GET_CURRENT_VOLUME:
                result(audibleHandler.getCurrentVolume())
                break;
            case Constants.SET_VOLUME:
                if let args = call.arguments as? Dictionary<String, Any>{
                    audibleHandler.setVolume(volume: (args["volume"] as? NSNumber)?.floatValue)
                }
                result(true)
                break;
            case Constants.GET_MAX_VOLUME:
                result(audibleHandler.getMaxVolume())
                break;
            default:
                result("Method not found")
            }
        }
        _currentVolumeHandler = CurrentVolumeStreamHandler(currentVolumeEventChannel)
        currentProfileEventChannel.setStreamHandler(CurrentProfileStreamHandler())
    }
    
    public func dispose(){
        channel.setMethodCallHandler(nil)
        currentProfileEventChannel.setStreamHandler(nil)
        currentVolumeEventChannel.setStreamHandler(nil)
    }
}
